File: src/workers/methods/letkf.js
import { periodicDist, gaspariCohn } from '../math';

export function buildLetkfPrecomputed(N, obsIndices, R_diag, localization) {
  const nobs = obsIndices.length;
  const localObsCount = new Int32Array(N);
  const tempIndices = [];
  const tempRInv = [];
  let maxLocalNobs = 0;

  for (let i = 0; i < N; i++) {
    let count = 0;
    for (let ob = 0; ob < nobs; ob++) {
      const dist = periodicDist(i, obsIndices[ob], N);
      const gloc = gaspariCohn(dist, localization);
      if (gloc > 1e-4) {
        tempIndices.push(ob);
        tempRInv.push(gloc / R_diag);
        count++;
      }
    }
    localObsCount[i] = count;
    if (count > maxLocalNobs) maxLocalNobs = count;
  }

  const localObsIndices = new Int32Array(N * maxLocalNobs);
  const localRInv = new Float64Array(N * maxLocalNobs);
  let srcIdx = 0;

  for (let i = 0; i < N; i++) {
    const count = localObsCount[i];
    for (let k = 0; k < count; k++) {
      localObsIndices[i * maxLocalNobs + k] = tempIndices[srcIdx + k];
      localRInv[i * maxLocalNobs + k] = tempRInv[srcIdx + k];
    }
    srcIdx += count;
  }

  return { localObsCount, localObsIndices, localRInv, maxLocalNobs };
}

export function updateLETKF(state, y, nobs, N, obsIndices, R_diag, localization, precomputed) {
  const M = state.ensembleSize;
  const ens = state.ensemble;
  const x_mean = state.x_mean;

  // Pre-compute perturbation matrix Xb (N x M)
  const Xb = new Array(N);
  for (let i = 0; i < N; i++) {
    const row = new Float64Array(M);
    const xm = x_mean[i];
    for (let j = 0; j < M; j++) row[j] = ens[j][i] - xm;
    Xb[i] = row;
  }

  const ens_new = new Array(M);
  for (let j = 0; j < M; j++) ens_new[j] = new Float64Array(N);
  
  const sqrtM1 = Math.sqrt(M - 1);
  const M1 = M - 1;
  const inv_sqrt_M = 1.0 / Math.sqrt(M);

  // Pre-allocated working buffers
  const Pa_tilde_inv = new Array(M);
  for (let r = 0; r < M; r++) Pa_tilde_inv[r] = new Float64Array(M);
  const L = new Array(M);
  for (let r = 0; r < M; r++) L[r] = new Float64Array(M);
  const b = new Float64Array(M);
  const y_temp = new Float64Array(M);
  const wa_mean = new Float64Array(M);
  const L_inv_T = new Array(M);
  for (let r = 0; r < M; r++) L_inv_T[r] = new Float64Array(M);
  const v_vec = new Float64Array(M);
  const u = new Float64Array(M);
  const y_sol = new Float64Array(M);

  const pre = precomputed || buildLetkfPrecomputed(N, obsIndices, R_diag, localization);
  const maxLoc = pre.maxLocalNobs;

  for (let i = 0; i < N; i++) {
    const l_nobs = pre.localObsCount[i];
    if (l_nobs === 0) {
      for (let j = 0; j < M; j++) ens_new[j][i] = ens[j][i];
      continue;
    }

    const locBase = i * maxLoc;

    // 1. Reset Pa_tilde_inv = (M-1) * I
    for (let r = 0; r < M; r++) {
      Pa_tilde_inv[r].fill(0);
      Pa_tilde_inv[r][r] = M1;
    }
    
    for (let o = 0; o < l_nobs; o++) {
      const oIdx = pre.localObsIndices[locBase + o];
      const obIdx = obsIndices[oIdx];
      const rInv = pre.localRInv[locBase + o];
      const Xb_ob = Xb[obIdx];

      for (let r = 0; r < M; r++) {
        const xr = Xb_ob[r] * rInv;
        const Pa_r = Pa_tilde_inv[r];
        for (let c = 0; c <= r; c++) {
          Pa_r[c] += xr * Xb_ob[c];
        }
      }
    }
    for (let r = 0; r < M; r++) {
      for (let c = 0; c < r; c++) {
        Pa_tilde_inv[c][r] = Pa_tilde_inv[r][c];
      }
    }

    // 2. Cholesky L * L^T = Pa_tilde_inv
    for (let r = 0; r < M; r++) L[r].fill(0);
    for (let r = 0; r < M; r++) {
      let sumDiag = 0;
      const L_r = L[r];
      for (let k = 0; k < r; k++) {
        const l_rk = L_r[k];
        sumDiag += l_rk * l_rk;
      }
      const l_rr = Math.sqrt(Math.max(1e-12, Pa_tilde_inv[r][r] - sumDiag));
      L_r[r] = l_rr;
      const inv_l_rr = 1.0 / l_rr;

      for (let r2 = r + 1; r2 < M; r2++) {
        let sum = 0;
        const L_r2 = L[r2];
        for (let k = 0; k < r; k++) sum += L_r2[k] * L_r[k];
        L_r2[r] = (Pa_tilde_inv[r2][r] - sum) * inv_l_rr;
      }
    }

    // 3. b = Y^T R^-1 (y - Hx_mean)
    b.fill(0);
    for (let o = 0; o < l_nobs; o++) {
      const oIdx = pre.localObsIndices[locBase + o];
      const obIdx = obsIndices[oIdx];
      const innov = (y[oIdx] - x_mean[obIdx]) * pre.localRInv[locBase + o];
      const Xb_ob = Xb[obIdx];
      for (let r = 0; r < M; r++) b[r] += Xb_ob[r] * innov;
    }

    // 4. Solve L * L^T * wa_mean = b
    for (let r = 0; r < M; r++) {
      let sum = 0;
      const L_r = L[r];
      for (let c = 0; c < r; c++) sum += L_r[c] * y_temp[c];
      y_temp[r] = (b[r] - sum) / L_r[r];
    }
    for (let r = M - 1; r >= 0; r--) {
      let sum = 0;
      for (let c = r + 1; c < M; c++) sum += L[c][r] * wa_mean[c];
      wa_mean[r] = (y_temp[r] - sum) / L[r][r];
    }

    // 5. L_inv_T: solve L * y = e_k for each k
    for (let k = 0; k < M; k++) {
      for (let r = 0; r < M; r++) {
        let sum = 0;
        const L_r = L[r];
        for (let c = 0; c < r; c++) sum += L_r[c] * y_sol[c];
        const rhs = r === k ? 1.0 : 0.0;
        y_sol[r] = (rhs - sum) / L_r[r];
      }
      for (let j = 0; j < M; j++) L_inv_T[k][j] = y_sol[j] * sqrtM1;
    }

    // 6. Householder transform
    for (let r = 0; r < M; r++) {
      let sum = 0;
      for (let c = 0; c < M; c++) sum += L_inv_T[c][r];
      v_vec[r] = sum * inv_sqrt_M;
    }
    
    let norm_v_sq = 0;
    for (let r = 0; r < M; r++) norm_v_sq += v_vec[r] * v_vec[r];
    const norm_v = Math.sqrt(norm_v_sq);
    
    for (let idx = 0; idx < M; idx++) u[idx] = v_vec[idx] - norm_v * inv_sqrt_M;
    let norm_u_sq = 0;
    for (let idx = 0; idx < M; idx++) norm_u_sq += u[idx] * u[idx];
    const norm_u = Math.sqrt(norm_u_sq);

    const Xb_i = Xb[i];
    const xm = x_mean[i];

    if (norm_u > 1e-12) {
      const inv_norm_u = 1.0 / norm_u;
      for (let idx = 0; idx < M; idx++) u[idx] *= inv_norm_u;

      for (let j = 0; j < M; j++) {
        let sum = 0;
        const u_j_2 = 2 * u[j];
        for (let k = 0; k < M; k++) {
          let dot_k = 0;
          const L_k = L_inv_T[k];
          for (let l = 0; l < M; l++) dot_k += L_k[l] * u[l];
          const wa_kj = L_k[j] - u_j_2 * dot_k;
          sum += Xb_i[k] * (wa_mean[k] + wa_kj);
        }
        ens_new[j][i] = xm + sum;
      }
    } else {
      for (let j = 0; j < M; j++) {
        let sum = 0;
        for (let k = 0; k < M; k++) {
          sum += Xb_i[k] * (wa_mean[k] + L_inv_T[k][j]);
        }
        ens_new[j][i] = xm + sum;
      }
    }
  }
  state.ensemble = ens_new;
}

File: crates/eduda_wasm/src/methods/letkf.rs
use crate::math::{periodic_dist, gaspari_cohn};

pub fn update_letkf(
    ensemble: &mut [f64], // M * N flat slice
    x_mean: &[f64],
    y: &[f64],
    obs_indices: &[usize],
    r_diag: f64,
    localization: f64,
    m: usize,
    n: usize,
) {
    let nobs = obs_indices.len();
    if nobs == 0 { return; }

    let sqrt_m1 = ((m - 1) as f64).sqrt();
    let m1_f64 = (m - 1) as f64;
    let inv_sqrt_m = 1.0 / (m as f64).sqrt();

    // Perturbation matrix Xb (N x M): Xb[i, j] = ensemble[j * N + i] - x_mean[i]
    let mut xb = vec![0.0; n * m];
    for i in 0..n {
        for j in 0..m {
            xb[i * m + j] = ensemble[j * n + i] - x_mean[i];
        }
    }

    let mut ens_new = vec![0.0; m * n];

    // Pre-allocated reusable working buffers
    let mut pa_tilde_inv = vec![0.0; m * m];
    let mut l_mat = vec![0.0; m * m];
    let mut b_vec = vec![0.0; m];
    let mut y_temp = vec![0.0; m];
    let mut wa_mean = vec![0.0; m];
    let mut l_inv_t = vec![0.0; m * m];
    let mut v_vec = vec![0.0; m];
    let mut u_vec = vec![0.0; m];
    let mut e_vec = vec![0.0; m];
    let mut y_sol = vec![0.0; m];

    let mut local_obs_idx = vec![0usize; nobs];
    let mut local_gloc = vec![0.0; nobs];

    for i in 0..n {
        let mut l_nobs = 0;
        for ob in 0..nobs {
            let dist = periodic_dist(i, obs_indices[ob], n);
            let gloc = gaspari_cohn(dist, localization);
            if gloc > 1e-4 {
                local_obs_idx[l_nobs] = ob;
                local_gloc[l_nobs] = gloc;
                l_nobs += 1;
            }
        }

        if l_nobs == 0 {
            for j in 0..m {
                ens_new[j * n + i] = ensemble[j * n + i];
            }
            continue;
        }

        // Reset Pa_tilde_inv = (M-1) * I
        pa_tilde_inv.fill(0.0);
        for r in 0..m {
            pa_tilde_inv[r * m + r] = m1_f64;
        }

        for r in 0..m {
            for c in 0..=r {
                let mut sum = 0.0;
                for o in 0..l_nobs {
                    let ob_idx = obs_indices[local_obs_idx[o]];
                    let r_loc_inv = local_gloc[o] / r_diag;
                    sum += xb[ob_idx * m + r] * r_loc_inv * xb[ob_idx * m + c];
                }
                pa_tilde_inv[r * m + c] += sum;
                if r != c {
                    pa_tilde_inv[c * m + r] = pa_tilde_inv[r * m + c];
                }
            }
        }

        // Cholesky L * L^T = Pa_tilde_inv
        l_mat.fill(0.0);
        for r in 0..m {
            for c in 0..=r {
                let mut sum = 0.0;
                for k in 0..c {
                    sum += l_mat[r * m + k] * l_mat[c * m + k];
                }
                if r == c {
                    let val = pa_tilde_inv[r * m + r] - sum;
                    l_mat[r * m + c] = val.max(1e-12).sqrt();
                } else {
                    l_mat[r * m + c] = (pa_tilde_inv[r * m + c] - sum) / l_mat[c * m + c];
                }
            }
        }

        // b = Y^T R^-1 (y - Hx_mean)
        for r in 0..m {
            let mut sum = 0.0;
            for o in 0..l_nobs {
                let o_idx = local_obs_idx[o];
                let ob_idx = obs_indices[o_idx];
                let innov = y[o_idx] - x_mean[ob_idx];
                sum += xb[ob_idx * m + r] * (local_gloc[o] / r_diag) * innov;
            }
            b_vec[r] = sum;
        }

        // Solve L * L^T * wa_mean = b
        for r in 0..m {
            let mut sum = 0.0;
            for c in 0..r {
                sum += l_mat[r * m + c] * y_temp[c];
            }
            y_temp[r] = (b_vec[r] - sum) / l_mat[r * m + r];
        }
        for r in (0..m).rev() {
            let mut sum = 0.0;
            for c in (r + 1)..m {
                sum += l_mat[c * m + r] * wa_mean[c];
            }
            wa_mean[r] = (y_temp[r] - sum) / l_mat[r * m + r];
        }

        // L_inv_T: solve L * y = e_k for each k, scale by sqrtM1
        for k in 0..m {
            e_vec.fill(0.0);
            e_vec[k] = 1.0;
            for r in 0..m {
                let mut sum = 0.0;
                for c in 0..r {
                    sum += l_mat[r * m + c] * y_sol[c];
                }
                y_sol[r] = (e_vec[r] - sum) / l_mat[r * m + r];
            }
            for j in 0..m {
                l_inv_t[k * m + j] = y_sol[j] * sqrt_m1;
            }
        }

        // Householder matrix Q
        for r in 0..m {
            let mut sum = 0.0;
            for c in 0..m {
                sum += l_inv_t[c * m + r] * inv_sqrt_m;
            }
            v_vec[r] = sum;
        }

        let mut norm_v_sq = 0.0;
        for r in 0..m {
            norm_v_sq += v_vec[r] * v_vec[r];
        }
        let norm_v = norm_v_sq.sqrt();

        for idx in 0..m {
            u_vec[idx] = v_vec[idx] - norm_v * inv_sqrt_m;
        }
        let mut norm_u_sq = 0.0;
        for idx in 0..m {
            norm_u_sq += u_vec[idx] * u_vec[idx];
        }
        let norm_u = norm_u_sq.sqrt();

        if norm_u > 1e-12 {
            let inv_norm_u = 1.0 / norm_u;
            for idx in 0..m {
                u_vec[idx] *= inv_norm_u;
            }
            for j in 0..m {
                let mut sum = 0.0;
                for k in 0..m {
                    let mut dot_k = 0.0;
                    for l in 0..m {
                        dot_k += l_inv_t[k * m + l] * u_vec[l];
                    }
                    let wa_kj = l_inv_t[k * m + j] - 2.0 * u_vec[j] * dot_k;
                    sum += xb[i * m + k] * (wa_mean[k] + wa_kj);
                }
                ens_new[j * n + i] = x_mean[i] + sum;
            }
        } else {
            for j in 0..m {
                let mut sum = 0.0;
                for k in 0..m {
                    sum += xb[i * m + k] * (wa_mean[k] + l_inv_t[k * m + j]);
                }
                ens_new[j * n + i] = x_mean[i] + sum;
            }
        }
    }

    ensemble.copy_from_slice(&ens_new);
}
